pakyow-support/lib/pakyow/support/indifferentize.rb

```diff
@@ -1,19 +1,112 @@
 module Pakyow
   module Support
-    class IndifferentHash < Hash
-      def [](key)
-        super(key.to_s)
+    # Creates a Hash-like object can access stored data with symbol or
+    #   string keys.
+    #
+    # The original hash is converted to symbol keys, which means that a hash
+    #   that originally contains a symbol and string key with the same frozen
+    #   string value will conflict. It is not guaranteed which value will
+    #   be saved.
+    #
+    # @example
+    #   { test: 'test1', 'test' => 'test2' } => { test: 'test2' }
+    #
+    class IndifferentHash < SimpleDelegator
+      class << self
+        def deep(hash)
+          pairs = hash.each_pair.map do |key, value|
+            if value.is_a? Hash
+              value = deep(value)
+            end
+            [key, value]
+          end
+
+          self.new(Hash[pairs])
+        end
+
+        def indifferent_key_method(*methods)
+          methods.each do |name|
+            define_method(name) do |key = nil, *args, &block|
+              key = convert_key(key)
+              internal_hash.public_send(name, key, *args, &block)
+            end
+          end
+        end
+
+        def indifferent_multi_key_method(*methods)
+          methods.each do |name|
+            define_method(name) do |*keys, &block|
+              keys = keys.map do |key|
+                convert_key(key)
+              end
+              internal_hash.public_send(name, *keys, &block)
+            end
+          end
+        end
+
+        def indifferentize_method(*methods)
+          methods.each do |name|
+            define_method(name) do |*args, &block|
+              hash = internal_hash.public_send(name, *args, &block)
+              self.class.new(hash)
+            end
+          end
+        end
+
+        def indifferentize_update_method(*methods)
+          methods.each do |name|
+            define_method(name) do |hash, &block|
+              hash = symbolize_keys(hash)
+              internal_hash.public_send(name, hash, &block)
+            end
+          end
+        end
       end
 
-      def []=(key, value)
-        super(key.to_s, value)
+      def initialize(hash)
+        self.internal_hash = hash
+      end
+
+      indifferent_key_method :[], :[]=, :default, :delete, :dig, :fetch, :has_key?, :key?, :include?, :member?, :store
```

Comment:
How far should we take the documentation on this object? I'd like for someone to understand how to use this object without having to find and browse Ruby's `Hash` docs, but maintaining our own version of core docs seems kind of lame as well. Wish there was a way to link to Ruby core methods (e.g. `@see Hash#key?`) but AFAIK this is only possible for internal docs.

Thoughts?